### PR TITLE
feat/TV-58-implement-createRoleUseCase

### PR DESCRIPTION
**Pull request by frenchMontajes**
## PR Question
**Do you have specific questions or issues you would like to address with the reviewer?**
- None

## Description
- created createRoleUseCase

## PR/Code Author Checklist
**Note:** please write "x" within the brackets if item applies e.g. [x]

- [x] I set the PR title based on the following format: [{task_type}] {task_id}: {task_name} . e.g. [feat] TV-X: add-new-functionality (camelCase)

- [x] I set the branch name based on the following format: {branch_type}/{task_id}-{task_name}. e.g feat/LM-X-add-new-functionality

- [x] I set KuroXI as the reviewer of this PR

- [x] I set myself as an assignees of this PR

- [x] I set the correct labels of this PR

- [x] All new and existing tests passed

- [x] I did not skip/commented-out/removed any existing tests

- [x] I have performed a self-review of my code

### CODE REVIEW

File: src/modules/user/src/useCases/role/createRoleUseCase.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+	public async execute(request: CreateRoleDTO) {
+		const roleOrFailure = RoleFactory.create({
+			id: uuid(),
+			name: request.name,
+			permissions: request.permissions,
+			color: request.color,
+			createdAt: new Date(),
+			updatedAt: null,
+			isDeleted: false,
+			deletedAt: null,
```
Comment (Arvicss) [inline src/modules/user/src/useCases/role/createRoleUseCase.ts:32]: remove createdAt, updatedAt, isDeleted, deletedAt

File: src/modules/user/src/useCases/role/createRoleUseCase.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+		});
+
```
Comment (Arvicss) [inline src/modules/user/src/useCases/role/createRoleUseCase.ts:34]: remove empty line

File: src/modules/user/src/useCases/role/createRoleUseCase.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+		if (roleOrFailure.isFailure) {
+			throw new BadRequestError(`failed to create role ${roleOrFailure.getErrorMessage}`);
+		}
+		const role = await this._roleRepository.createRole(roleOrFailure.getValue());
```
Comment (Arvicss) [inline src/modules/user/src/useCases/role/createRoleUseCase.ts:38]: add empty line before this variable

File: src/modules/user/src/useCases/role/createRoleUseCase.ts [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
```
Comment (Arvicss) [inline src/modules/user/src/useCases/role/createRoleUseCase.ts:39]: remove empty line

File: src/modules/user/tests/infrastructure/http/routes/role/createRole.test.ts [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+import { createCallerFactory } from "@/shared/infrastructure/trpc";
+import { faker } from "@faker-js/faker";
+import type { TRPCError } from "@trpc/server";
+
+describe("createRoleEndpoint", () => {
+	let roleRepository: IRoleRepository;
+	beforeAll(() => {
+		roleRepository = new RoleRepository();
+	});
+	describe("User is authenticated", () => {
```
Comment (Arvicss) [inline src/modules/user/tests/infrastructure/http/routes/role/createRole.test.ts:19]: fix spacing add empty line in between

File: src/modules/user/tests/infrastructure/http/routes/role/createRole.test.ts [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+				},
+				db: db,
+			});
+
+			const roleId = await caller.createRole(request);
+			const createRole = await roleRepository.getRoleById(roleId);
+			expect(createRole).not.toBeNull();
+			expect(createRole!.nameValue).toBe(request.name);
+			expect(createRole!.permissionsValue).toBe(request.permissions);
+			expect(createRole!.color).toBe(request.color);
```
Comment (Arvicss) [inline src/modules/user/tests/infrastructure/http/routes/role/createRole.test.ts:47]: Remove this assertion. Out of test scope.

File: src/modules/user/tests/infrastructure/http/routes/role/createRole.test.ts [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+				session: {
+					user: {
+						id: seededUser.id,
+					},
+					expires: new Date().toString(),
+				},
+				db: db,
+			});
+
+			const roleId = await caller.createRole(request);
```
Comment (Arvicss) [inline src/modules/user/tests/infrastructure/http/routes/role/createRole.test.ts:42]: assert that roleId is not null

File: src/modules/user/src/useCases/role/createRoleUseCase.ts [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+	public constructor(
+		roleRepository = new RoleRepository(),
+		roleAuditLogService = new RoleAuditLogService(),
+	) {
+		this._roleRepository = roleRepository;
+		this._roleAuditLogService = roleAuditLogService;
+	}
+
+	public async execute(request: CreateRoleDTO) {
```
Comment (Arvicss) [inline src/modules/user/src/useCases/role/createRoleUseCase.ts:23]: add return type

File: src/modules/user/tests/useCases/role/createRoleUseCase.test.ts [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+} from "@/modules/user/src/repositories/roleAuditLogRepository";
+import { RoleRepository } from "@/modules/user/src/repositories/roleRepository";
+import { Permissions } from "@/modules/user/src/shared/permissions";
+import { CreateRoleUseCase } from "@/modules/user/src/useCases/role/createRoleUseCase";
+import { faker } from "@faker-js/faker";
+import { seedUser } from "../../utils/user/seedUser";
+
+describe("CreateRoleUseCase", () => {
+	let createRoleUseCase: CreateRoleUseCase;
+	let roleRepository: RoleRepository;
```
Comment (Arvicss) [inline src/modules/user/tests/useCases/role/createRoleUseCase.test.ts:14]: use `IRoleRepository`

File: src/modules/user/tests/useCases/role/createRoleUseCase.test.ts [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+		};
+
+		const result = await createRoleUseCase.execute(request);
+		const createdRole = await roleRepository.getRoleById(result);
+
+		expect(createdRole).not.toBeNull();
+		expect(createdRole!.nameValue).toBe(request!.name);
+		expect(createdRole!.color).toBe(request!.color);
+		expect(createdRole!.permissionsValue).toBe(request!.permissions);
+
```
Comment (Arvicss) [inline src/modules/user/tests/useCases/role/createRoleUseCase.test.ts:40]: assert the following:

`isDeleted`
`deletedAt`
`updatedAt`

**Review by Arvicss — CHANGES_REQUESTED** [review]
Rebase Branch

**Review by KuroXI — APPROVED** [review]

**Pull request merged**